### PR TITLE
feat(tags): Add common_tags() helper for standardized resource tagging

### LINKED ISSUE #16: [BENCH t01 gemini-3-flash-preview-c] feat(tags): Add common_tags() helper for standardized resource tagging
### Objective

feat(tags): Add common_tags() helper for standardized resource tagging

### Acceptance criteria

- [ ] Implementation matches all behaviors described in the task body (see Notes)
- [ ] All named tests pass the verification command
- [ ] No dependencies added unless absolutely required

### Out-of-scope / non-goals

- Do NOT modify files beyond the expected set below
- Do NOT add third-party dependencies unless required by the task body
- Do NOT refactor unrelated code in the touched files

### Files expected to change

- `infiquetra_aws_infra/tagging.py`
- `tests/test_tagging.py`

### Tests to add or update

- See the task body for the named test cases

### Verification

```bash
cd ~/workspace/infiquetra/infiquetra-aws-infra
# Language-appropriate test command (flutter test / pytest / etc.)
```

## Task

Add `common_tags(env: str, team: str, project: str) -> dict[str, str]` to `infiquetra_aws_infra/tagging.py`.

## Requirements

- Return a dict with the keys: `Environment`, `Team`, `Project`, `ManagedBy`.
- `ManagedBy` must always be `"CDK"`.
- `env` must be one of `"dev"`, `"staging"`, `"prod"` — raise `ValueError` on others.
- Strip/normalize trailing whitespace on all inputs.
- No external dependencies.

## Examples

- `common_tags("dev", "platform", "auth")` → `{"Environment": "dev", "Team": "platform", "Project": "auth", "ManagedBy": "CDK"}`
- `common_tags("test", "x", "y")` → raises `ValueError`
- `common_tags("dev  ", " platform ", "auth")` → normalized keys

## Tests

Add `tests/test_tagging.py` with pytest: valid dev/staging/prod, ValueError for bad env, whitespace normalization, all keys present.

Scope: small (<40 lines).

### Context library links

- Benchmark task id: t01

### PR DESCRIPTION
**Pull request by namredips**
## Linked card

Closes #16

## What changed

- Added `infiquetra_aws_infra/tagging.py` with `common_tags(env, team, project) -> dict[str, str]`
  - Normalizes all three inputs via `.strip()`
  - Validates env against {dev, staging, prod}, raises `ValueError` on others
  - Returns keys: Environment, Team, Project, ManagedBy=CDK
- Added `tests/test_tagging.py` with 6 pytest cases covering valid envs, invalid env, whitespace normalization, and required keys

## How verified

```
cd ~/workspace/infiquetra/infiquetra-aws-infra
PYTHONPATH=. pytest tests/test_tagging.py -v
```
Result: 6 passed in 0.02s

`ruff check` on both files: all checks passed

## Acceptance criteria coverage

- AC 1 (implementation matches all behaviors): ✓ addressed in `infiquetra_aws_infra/tagging.py::common_tags`
- AC 2 (all named tests pass): ✓ addressed in `tests/test_tagging.py` — 6 tests, all passing
- AC 3 (no new dependencies): ✓ addressed — pure Python stdlib, existing pytest from dev deps

## Non-goals acknowledged

- Do NOT modify files beyond the expected set: not violated — only `tagging.py` and `test_tagging.py` changed
- Do NOT add third-party dependencies: not violated — no new deps introduced
- Do NOT refactor unrelated code: not violated — no refactors outside scoped files

## Notes

- Python 3.13 required by pyproject.toml but not available on the build VM; tests run via `PYTHONPATH=. pytest` to avoid editable install
- ManagedBy is hard-coded to "CDK" per the issue spec

### Coverage

- AC 1 (verbatim): ✓ addressed in `infiquetra_aws_infra/tagging.py::common_tags` — strips inputs, validates env, returns required keys with ManagedBy=CDK
- AC 2 (verbatim): ✓ addressed in `tests/test_tagging.py` — 6 tests covering valid envs, ValueError for invalid env, whitespace normalization, required keys present
- AC 3 (verbatim): ✓ addressed — pure Python helper, pytest from existing dev dependencies

### CODE REVIEW

**Comment by namredips** [comment]
## ✅ PR review round 1 — MERGE-READY

**Overall:** 8.90/10 · **HEAD:** `ce3fb9bd`

_Per-reviewer scores: ac_verification 0.0 · code_quality 7.9 · plan_fidelity 9.7 · robustness 8.6 · security_code 8.6 · python_types 9.0 · test_presence 9.6_

### Reviewer verdicts

<details open>
<summary>🔄 <strong>ac_verification</strong> — 0.00/10</summary>

[reviewer LLM failed after retries — verdict is synthetic, excluded from aggregate. Error: Command '['/Users/jefcox/.local/bin/hermes', 'chat', '-q', '[SYSTEM: You are the `ac_verification` reviewer, evaluating an **ACTUAL PULL\nREQUEST** — code that exists. The agent has implemented the approved\nplan and pushed commits. Your job is to judge whether the code is\nmerge-ready from your rubric\'s perspective.\n\nYou receive the card as structured fields:\n\n**Objective:** feat(tags): Add common_tags() helper for standardized resource tagging\n\n**Acceptance criteria:**\n- [ ] Implementation matches all behaviors described in the task body (see Notes)\n- [ ] All named tests pass the verification command\n- [ ] No dependencies added unless absolutely required\n\n**Out-of-scope / non-goals:**\n- Do NOT modify files beyond the expected set below\n- Do NOT add third-party dependencies unless required by the task body\n- Do NOT refactor unrelated code in the touched files\n\n**Files expected to change:**\n- `infiquetra_aws_infra/tagging.py`\n- `tests/test_tagging.py`\n\n**Tests to add/update:**\n- See the task body for the named test cases\n\n**Verification:**\n```bash\ncd ~/workspace/infiquetra/infiquetra-aws-infra\n# Language-appropriate test command (flutter test / pytest / etc.)\n```\n\n**Notes / conventions:**\n## Task\n\nAdd `common_tags(env: str, team: str, project: str) -> dict[str, str]` to `infiquetra_aws_infra/tagging.py`.\n\n## Requirements\n\n- Return a dict with the keys: `Environment`, `Team`, `Project`, `ManagedBy`.\n- `ManagedBy` must always be `"CDK"`.\n- `env` must be one of `"dev"`, `"staging"`, `"prod"` — raise `ValueError` on others.\n- Strip/normalize trailing whitespace on all inputs.\n- No external dependencies.\n\n## Examples\n\n- `common_tags("dev", "platform", "auth")` → `{"Environment": "dev", "Team": "platform", "Project": "auth", "ManagedBy": "CDK"}`\n- `common_tags("test", "x", "y")` → raises `ValueError`\n- `common_tags("dev  ", " platform ", "auth")` → normalized keys\n\n## Tests\n\nAdd `tests/test_tagging.py` with pytest: valid dev/staging/prod, ValueError for bad env, whitespace normalization, all keys present.\n\nScope: small (<40 lines).\n\n**Context library links:** - Benchmark task id: t01\n\n\n\n\n\n\nAnd the **approved plan** the agent was supposed to implement:\n\n## Files to modify\n- `infiquetra_aws_infra/tagging.py`\n  - Verified with `gh api repos/infiquetra/infiquetra-aws-infra/contents/infiquetra_aws_infra --jq \'.[].path\'`: the package directory exists and currently contains `__init__.py`, `organization_stack.py`, and `sso_stack.py`, but no `tagging.py` yet.\n  - Add the new `common_tags(env: str, team: str, project: str) -> dict[str, str]` helper in this new module, following the repo’s existing typed Python style visible in `organization_stack.py` and `sso_stack.py`.\n\n- `tests/test_tagging.py`\n  - Verified with `gh api repos/infiquetra/infiquetra-aws-infra/contents/tests`: the `tests/` directory is not present yet, so this test module will be a new file and will require creating its parent directory.\n  - Verified with `gh api repos/infiquetra/infiquetra-aws-infra/contents/pyproject.toml -H \'Accept: application/vnd.github.raw+json\'`: `[tool.pytest.ini_options]` sets `testpaths = ["tests"]` and `python_files = ["test_*.py", "*_test.py"]`, so `tests/test_tagging.py` is the correct pytest location and naming convention.\n  - Add pytest coverage for the new `common_tags(...)` helper: valid `dev`/`staging`/`prod` cases, invalid env `ValueError`, whitespace normalization, and all required keys present.\n\n## Approach\n1. Create `infiquetra_aws_infra/tagging.py` and implement `common_tags(env: str, team: str, project: str) -> dict[str, str]` as a small standalone helper with no new imports beyond the standard library or none at all.\n2. In `common_tags`, normalize all three string inputs before any branching by applying `.strip()` to `env`, `team`, and `project`, so the helper satisfies the example where values like `"dev  "` and `" platform "` are returned normalized.\n3. Validate the normalized environment value against the allowed set `{"dev", "staging", "prod"}`.\n   - Only if normalization succeeds and the normalized `env` is one of those three values, proceed to construct the return dictionary.\n   - If normalization succeeds but the normalized `env` is any other value, the function must raise `ValueError` immediately rather than returning a partial/default tag set.\n4. Return a dict whose exact keys are `Environment`, `Team`, `Project`, and `ManagedBy`, wiring the normalized values into `Environment`, `Team`, and `Project`, and hard-coding `ManagedBy` to `"CDK"`.\n5. Create `tests/test_tagging.py` with pytest tests targeted at the new helper.\n   - Add one test that covers the happy-path environments `dev`, `staging`, and `prod`, asserting the returned dict values match the normalized inputs and that `ManagedBy == "CDK"`.\n   - Add one test that passes an unsupported env such as `"test"` and asserts `common_tags(...)` raises `ValueError`.\n   - Add one test for whitespace normalization using the Notes example shape (for example `"dev  "`, `" platform "`, `"auth"`) and assert the returned values are stripped.\n   - Add one test that asserts the returned dict has exactly the required keys `Environment`, `Team`, `Project`, and `ManagedBy`.\n6. Verify the change with pytest from the repo root configured in the card.\n   - Only if the test command runs and all new/updated tests pass, the task is complete.\n   - If the test command fails, the task must fail and the agent should fix only the implementation/tests within the scoped files rather than broadening the change.\n\n## Acceptance criteria coverage\n- AC 1: covered by steps 2, 3, and 4 — the `common_tags(env: str, team: str, project: str) -> dict[str, str]` helper in `infiquetra_aws_infra/tagging.py` strips the three inputs with `.strip()`, validates the normalized `env` against `{"dev", "staging", "prod"}`, raises `ValueError` for any other normalized value, and returns the exact required keys `Environment`, `Team`, `Project`, and `ManagedBy` with `ManagedBy` fixed to `"CDK"`.\n- AC 2: covered by steps 5 and 6 — `tests/test_tagging.py` adds the named pytest coverage for valid `dev`/`staging`/`prod`, bad env `ValueError`, whitespace normalization, and all keys present, and step 6 runs pytest from the repository root specified in the card.\n- AC 3: covered by step 1 and confirmed by the fetched `pyproject.toml` — `common_tags(...)` is implemented as a small pure-Python helper in `infiquetra_aws_infra/tagging.py`, and `tests/test_tagging.py` uses the already-configured pytest toolchain from `[dependency-groups].dev` / `[tool.pytest.ini_options]`, so no new dependencies are introduced.\n\n## Out-of-scope acknowledged\n- Do NOT modify files beyond the expected set below — acknowledged; the plan changes only `infiquetra_aws_infra/tagging.py` and `tests/test_tagging.py`.\n- Do NOT add third-party dependencies unless required by the task body — acknowledged; the plan uses existing Python and pytest setup only.\n- Do NOT refactor unrelated code in the touched files — acknowledged; the plan adds one helper and one focused test module without unrelated cleanup or refactors.\n\n## Tests / verification\n- New test file: `tests/test_tagging.py`\n- Recommended test functions:\n  - `test_common_tags_accepts_valid_environments`\n  - `test_common_tags_rejects_invalid_environment`\n  - `test_common_tags_strips_whitespace_from_inputs`\n  - `test_common_tags_returns_all_required_keys`\n- Verification commands:\n  - `cd ~/workspace/infiquetra/infiquetra-aws-infra`\n  - `pytest tests/test_tagging.py`\n## Files to modify\n- `infiquetra_aws_infra/tagging.py`\n  - Verified with `gh api repos/infiquetra/infiquetra-aws-infra/contents/infiquetra_aws_infra --jq \'.[].path\'`: the package directory exists and currently contains `__init__.py`, `organization_stack.py`, and `sso_stack.py`, but no `tagging.py` yet.\n  - Add the new `common_tags(env: str, team: str, project: str) -> dict[str, str]` helper in this new module, following the repo’s existing typed Python style visible in `organization_stack.py` and `sso_stack.py`.\n\n- `tests/test_tagging.py`\n  - Verified with `gh api repos/infiquetra/infiquetra-aws-infra/contents/tests`: the `tests/` directory is not present yet, so this test module will be a new file and will require creating its parent directory.\n  - Verified with `gh api repos/infiquetra/infiquetra-aws-infra/contents/pyproject.toml -H \'Accept: application/vnd.github.raw+json\'`: `[tool.pytest.ini_options]` sets `testpaths = ["tests"]` and `python_files = ["test_*.py", "*_test.py"]`, so `tests/test_tagging.py` is the correct pytest location and naming convention.\n  - Add pytest coverage for the new `common_tags(...)` helper: valid `dev`/`staging`/`prod` cases, invalid env `ValueError`, whitespace normalization, and all required keys present.\n\n## Approach\n1. Create `infiquetra_aws_infra/tagging.py` and implement `common_tags(env: str, team: str, project: str) -> dict[str, str]` as a small standalone helper with no new imports beyond the standard library or none at all.\n2. In `common_tags`, normalize all three string inputs before any branching by applying `.strip()` to `env`, `team`, and `project`, so the helper satisfies the example where values like `"dev  "` and `" platform "` are returned normalized.\n3. Validate the normalized environment value against the allowed set `{"dev", "staging", "prod"}`.\n   - Only if normalization succeeds and the normalized `env` is one of those three values, proceed to construct the return dictionary.\n   - If normalization succeeds but the normalized `env` is any other value, the function must raise `ValueError` immediately rather than returning a partial/default tag set.\n4. Return a dict whose exact keys are `Environment`, `Team`, `Project`, and `ManagedBy`, wiring the normalized values into `Environment`, `Team`, and `Project`, and hard-coding `ManagedBy` to `"CDK"`.\n5. Create `tests/test_tagging.py` with pytest tests targeted at the new helper.\n   - Add one test that covers the happy-path environments `dev`, `staging`, and `prod`, asserting the returned dict values match the normalized inputs and that `ManagedBy == "CDK"`.\n   - Add one test that passes an unsupported env such as `"test"` and asserts `common_tags(...)` raises `ValueError`.\n   - Add one test for whitespace normalization using the Notes example shape (for example `"dev  "`, `" platform "`, `"auth"`) and assert the returned values are stripped.\n   - Add one test that asserts the returned dict has exactly the required keys `Environment`, `Team`, `Project`, and `ManagedBy`.\n6. Verify the change with pytest from the repo root configured in the card.\n   - Only if the test command runs and all new/updated tests pass, the task is complete.\n   - If the test command fails, the task must fail and the agent should fix only the implementation/tests within the scoped files rather than broadening the change.\n\n## Acceptance criteria coverage\n- AC 1: covered by steps 2, 3, and 4 — the `common_tags(env: str, team: str, project: str) -> dict[str, str]` helper in `infiquetra_aws_infra/tagging.py` strips the three inputs with `.strip()`, validates the normalized `env` against `{"dev", "staging", "prod"}`, raises `ValueError` for any other normalized value, and returns the exact required keys `Environment`, `Team`, `Project`, and `ManagedBy` with `ManagedBy` fixed to `"CDK"`.\n- AC 2: covered by steps 5 and 6 — `tests/test_tagging.py` adds the named pytest coverage for valid `dev`/`staging`/`prod`, bad env `ValueError`, whitespace normalization, and all keys present, and step 6 runs pytest from the repository root specified in the card.\n- AC 3: covered by step 1 and confirmed by the fetched `pyproject.toml` — `common_tags(...)` is implemented as a small pure-Python helper in `infiquetra_aws_infra/tagging.py`, and `tests/test_tagging.py` uses the already-configured pytest toolchain from `[dependency-groups].dev` / `[tool.pytest.ini_options]`, so no new dependencies are introduced.\n\n## Out-of-scope acknowledged\n- Do NOT modify files beyond the expected set below — acknowledged; the plan changes only `infiquetra_aws_infra/tagging.py` and `tests/test_tagging.py`.\n- Do NOT add third-party dependencies unless required by the task body — acknowledged; the plan uses existing Python and pytest setup only.\n- Do NOT refactor unrelated code in the touched files — acknowledged; the plan adds one helper and one focused test module without unrelated cleanup or refactors.\n\n## Tests / verification\n- New test file: `tests/test_tagging.py`\n- Recommended test functions:\n  - `test_common_tags_accepts_valid_environments`\n  - `test_common_tags_rejects_invalid_environment`\n  - `test_common_tags_strips_whitespace_from_inputs`\n  - `test_common_tags_returns_all_required_keys`\n- Verification commands:\n  - `cd ~/workspace/infiquetra/infiquetra-aws-infra`\n  - `pytest tests/test_tagging.py`\n\nAnd the **PR diff** the agent produced:\n\n```diff\ndiff --git a/infiquetra_aws_infra/tagging.py b/infiquetra_aws_infra/tagging.py\nnew file mode 100644\nindex 0000000..f2c9c40\n--- /dev/null\n+++ b/infiquetra_aws_infra/tagging.py\n@@ -0,0 +1,34 @@\n+"""Standardized resource tagging helpers for AWS CDK stacks."""\n+\n+\n+def common_tags(env: str, team: str, project: str) -> dict[str, str]:\n+    """\n+    Return a standardized tag dict for AWS resources.\n+\n+    Args:\n+        env: Deployment environment. Must be one of "dev", "staging", "prod".\n+        team: Team responsible for the resource.\n+        project: Project name.\n+\n+    Returns:\n+        Dict with keys: Environment, Team, Project, ManagedBy.\n+\n+    Raises:\n+        ValueError: If env is not one of "dev", "staging", "prod".\n+    """\n+    env_norm = env.strip()\n+    team_norm = team.strip()\n+    project_norm = project.strip()\n+\n+    valid_envs = {"dev", "staging", "prod"}\n+    if env_norm not in valid_envs:\n+        raise ValueError(\n+            f"Invalid env {env_norm!r}: must be one of {sorted(valid_envs)}"\n+        )\n+\n+    return {\n+        "Environment": env_norm,\n+        "Team": team_norm,\n+        "Project": project_norm,\n+        "ManagedBy": "CDK",\n+    }\ndiff --git a/tests/test_tagging.py b/tests/test_tagging.py\nnew file mode 100644\nindex 0000000..615d204\n--- /dev/null\n+++ b/tests/test_tagging.py\n@@ -0,0 +1,35 @@\n+"""Tests for tagging helpers."""\n+\n+import pytest\n+\n+from infiquetra_aws_infra.tagging import common_tags\n+\n+\n+@pytest.mark.parametrize("env", ["dev", "staging", "prod"])\n+def test_common_tags_accepts_valid_environments(env: str) -> None:\n+    """Valid environments produce a tag dict with correct values."""\n+    tags = common_tags(env, "platform", "auth")\n+    assert tags["Environment"] == env\n+    assert tags["Team"] == "platform"\n+    assert tags["Project"] == "auth"\n+    assert tags["ManagedBy"] == "CDK"\n+\n+\n+def test_common_tags_rejects_invalid_environment() -> None:\n+    """Non-dev/staging/prod env raises ValueError."""\n+    with pytest.raises(ValueError, match="Invalid env"):\n+        common_tags("test", "x", "y")\n+\n+\n+def test_common_tags_strips_whitespace_from_inputs() -> None:\n+    """Input strings are normalized via strip()."""\n+    tags = common_tags("dev  ", " platform ", "auth")\n+    assert tags["Environment"] == "dev"\n+    assert tags["Team"] == "platform"\n+    assert tags["Project"] == "auth"\n+\n+\n+def test_common_tags_returns_all_required_keys() -> None:\n+    """Returned dict has exactly Environment, Team, Project, ManagedBy."""\n+    tags = common_tags("prod", "sre", "platform")\n+    assert set(tags.keys()) == {"Environment", "Team", "Project", "ManagedBy"}\n\n```\n\n## How to evaluate\n\n**Step 1 — Mechanical AC coverage check.**\nFor each acceptance criterion listed in the card, classify the diff\'s\ntreatment:\n- ADDRESSED — the diff\'s code produces the observable behavior this AC\n  demands (and a test in the diff proves it, or the card said "no\n  tests required" with a rationale you can verify).\n- UNADDRESSED — the diff doesn\'t produce the behavior; agent missed\n  or skipped this AC.\n- PARTIALLY — the diff covers the AC for the common case but a named\n  edge case isn\'t handled.\n\n**Step 2 — Non-goals check.**\nVerify the diff does NOT do any work on any non-goal. Scope creep\n(diff modifies files the card explicitly said to avoid, adds features\nthe non-goals forbid) is grounds for REVISE or BLOCK.\n\n**Step 3 — Rubric-specific qualitative review.**\nApply your rubric below as a lens over the diff. Your rubric is the\nauthoritative guide to what YOU care about. Other reviewers cover\nother lenses — stay in your lane.\n\nYour rubric:\n\n---\nphases: [pr]\napplicability: always\n---\n# AC Verification lens (PR-phase)\n\nYour focus: **for each Acceptance Criterion in the card, does the\ndiff (+ its tests) produce the observable behavior the AC demands?**\n\nYou are the customer-facing reviewer. ACs are the contract between\nthe card author and the automation. Plan approval didn\'t matter if\nthe diff doesn\'t actually satisfy the ACs.\n\n## How this differs from `plan_fidelity`\n\n- `plan_fidelity`: did the agent implement what the plan said?\n- **You (`ac_verification`)**: does the diff satisfy what the card\'s\n  ACs demand?\n\nA diff can be plan-faithful but AC-incomplete (the plan had a gap\nthe reviewers missed). Conversely, a diff can deviate from the plan\nbut still satisfy every AC (agent discovered a better path). You\njudge AC outcomes, not plan-step completion.\n\n## How to evaluate\n\nFor **each AC** in the card (you have them in your system prompt),\ndo the following:\n\n1. **Identify the observable.** What would a human verifier look for\n   to confirm this AC holds? An exit code? A specific output? A\n   behavior under a named input?\n2. **Find the evidence in the diff.** Does the code contain the\n   behavior? Does a test in the diff prove the behavior?\n3. **Check the Verification command.** The card has an exact\n   command in its Verification section. Does the diff make that\n   command pass?\n4. **Record coverage.** Return `ac_coverage` in your JSON verdict:\n   each AC number → ADDRESSED | PARTIALLY | UNADDRESSED.\n\n## What counts as "addressed"\n\n- **ADDRESSED**: the behavior the AC demands is visible in the\n  diff\'s code, AND a test in the diff proves it (OR the card said\n  "no new tests required" with a rationale you can verify holds).\n- **PARTIALLY**: the behavior is present but a named edge case the\n  AC mentions isn\'t handled (e.g., "handles malformed input" when\n  only well-formed input is in the code path).\n- **UNADDRESSED**: the AC\'s observable behavior is not in the diff,\n  or there\'s no way to verify it from what\'s here.\n\n## What to look for\n\n- **Happy-path coverage.** The core scenario each AC describes\n  actually works in the code.\n- **Test-AC linkage.** Each AC should trace to at least one assert,\n  ideally with a test name that references the AC\'s verb.\n- **Verification command green.** If the card\'s Verification command\n  fails on the diff, the AC isn\'t truly satisfied even if the code\n  "looks right."\n- **No AC silently dropped.** If the diff deliberately doesn\'t\n  address an AC, the PR description must acknowledge it (and it\n  should probably be a REVISE for re-scoping).\n\n## What NOT to flag\n\n- Code quality on correct code (`code_quality`).\n- Robustness beyond what the AC requires (`robustness`).\n- Security concerns unrelated to the ACs (`security_code`).\n- Plan deviation (that\'s `plan_fidelity`).\n\n## Scoring\n\n- **10**: Every AC is ADDRESSED. Tests (or existing-test citations)\n  map 1:1 to ACs. Verification command is green.\n- **9**: Every AC ADDRESSED; one test could be more specific.\n- **8**: All ACs at least PARTIALLY, one with a minor gap an agent\n  could close at PR time.\n- **7**: One AC UNADDRESSED OR one AC PARTIALLY with a non-trivial\n  gap.\n- **≤6**: Multiple ACs UNADDRESSED / PARTIALLY, or the diff is\n  fundamentally incomplete against the card.\n\n## REVISE criteria\n\nREVISE when you can name:\n  (a) the AC number + text,\n  (b) what\'s missing in the diff,\n  (c) what needs to be added — a specific test, a handling branch,\n      an error path.\n\n## BLOCK only for\n\n- Verification command fails on the diff with no plausible path\n  to green.\n- Half or more of the ACs are UNADDRESSED.\n\n## CARD: escalation\n\nIf an AC is genuinely unimplementable given the non-goals + files\nscope of the card, use `CARD:` prefix. The author either needs to\nrelax the non-goals, broaden `files_expected`, or soften the AC.\nDon\'t grind the implementer on an impossible constraint.\n\n\n## When the critique is about the CARD, not the code\n\nRare at PR-phase — most card-shape issues were caught at plan-phase.\nBut if a card-level inconsistency forces the code to be wrong (AC\ndemands behavior non-goals forbid, Verification command targets files\nthe diff can\'t produce, etc.), use the **`CARD:`** prefix on your\ncomment. The aggregator routes CARD:+BLOCK verdicts to\n`QUESTION_FOR_AUTHOR` so the card owner can resolve, not the\nimplementer.\n\nExamples:\n- ✅ `CARD: AC 3 demands accepting raw HTML input; non-goals forbid\n  sanitization; the diff ships an XSS surface. Author needs to\n  reconcile.`\n- ❌ `CARD: the code should use parameterized queries` ← code-level\n  critique, not card-level. Drop the CARD: prefix.\n\n## How to choose a verdict\n\n**PROCEED** when:\n- Every AC is ADDRESSED in Step 1.\n- No non-goal is violated.\n- Your rubric lens shows no code-level concern that rises to "this PR\n  shouldn\'t merge as-is".\n- Nits and minor suggestions can go in `comments` — they\'re\n  information, not blockers.\n\n**Calibrate your scoring carefully.** A diff with no concerns in your\ndomain is a 9-10 depending on whether the code actively demonstrates\ndefenses you can name. **A pure-data diff that touches nothing your\nrubric cares about should score 8-9 with a brief "no surface" note —\nNOT 9.8+.** Inflation of the mean on orthogonal diffs is the major\npathology we\'re correcting for; keep the top of the scale reserved\nfor actively-defended work.\n\n**REVISE** when:\n- One or more AC items are UNADDRESSED / PARTIALLY. Cite the AC\n  number and say what\'s missing.\n- A non-goal is violated. Cite the non-goal and the diff line.\n- Your rubric found a concrete code-level issue the agent should fix\n  before merge — name file + line + the fix.\n\nREVISE comments are CITATIONS + CONCRETE FIXES, not vague\nskepticism. "This could be cleaner" is PROCEED-with-comments.\n\n**BLOCK** for:\n- Code is fundamentally wrong from your lens: a hardcoded secret in\n  the diff, a data-loss migration with no rollback, an\n  unauthenticated endpoint to sensitive data. Something no PR-round\n  revision will fix without rearchitecting.\n- The verification command fails and the diff can\'t plausibly be\n  iterated to green.\n\nBLOCK is heavy — use sparingly. The aggregator interprets BLOCK as\n"halt the loop and escalate" (or, if CARD:-prefixed, route to the\nauthor).\n\n## Respect the division of labor\n\nYou are ONE reviewer. Other reviewers cover other lenses:\n- `code_quality` — craftsmanship (readability, placement, idiom)\n- `security_code` — exploits (injection, secrets, authz)\n- `robustness` — runtime behavior (edge cases, failure paths)\n- `plan_fidelity` — does the diff implement the approved plan\n- `ac_verification` — does the diff satisfy each AC\n- `test_presence` — tests adequate for ACs\n- Language reviewers (`python_types`, `dart_flutter`) — language\n  idioms + type safety\n- Policy reviewers — diff vs linked ADRs / standards / policies\n\nDon\'t grade the whole PR on your single axis — focus on what your\nrubric actually covers. Don\'t score low because the PR has a concern\nthat belongs to another reviewer.\n\n## Score calibration\n\n- **10**: Exemplary in your domain. Code actively demonstrates the\n  defenses / properties your rubric cares about.\n- **9**: Strong. No concerns in your domain, or one minor nit that\'s\n  comment-worthy not REVISE-worthy.\n- **8**: Safe / sound in your domain. Maybe the PR has nothing your\n  rubric materially cares about — a brief "no surface" note and an\n  8 is correct (avoid 9.8 score-inflation).\n- **7**: Works but has a specific concern in your domain that\n  should be tightened before merge.\n- **6 or below**: Concrete problem that makes this PR not\n  merge-ready from your lens.\n\n## Evidence rules — every REVISE claim must be backed by a citation\n\nBefore you claim anything about the code or tests, locate the specific\n`file:line` in the diff. **Claims without citations are invalid — the\naggregator will treat them as noise.** Example of a good claim:\n`lib/foo.dart:42 — missing null check before dereference`. Example\nof a bad claim: `the error handling is weak`.\n\n**Verify before flagging.** The most common failure mode of PR review\n(observed on smoke9 mimir#143 R2/R3, where three reviewers each\nhallucinated that test assertions were missing) is stopping after the\nfirst thing you see and concluding without reading the full context.\nBefore you say "the tests don\'t assert X":\n\n1. **Read the full assertion chain, not just the first line.** In\n   Dart, `throwsA(isA<E>().having(...).having(...).having(...))`\n   CHAINS multiple `.having(...)` calls on a single expectation. Look\n   at EVERY `.having` call before concluding what the test asserts —\n   don\'t stop at the first one you see.\n   In Python, `pytest.raises(..., match=r"...")` takes a regex —\n   read the full pattern. In JS, `expect(...).toThrow(...)` can\n   accept strings, regexes, or matcher objects — check which.\n2. **Grep the diff for X.** Before claiming "the diff doesn\'t have\n   X", search the `diff_text` for X. If X appears, you cannot make\n   that claim. Either rewrite your finding to describe a different\n   gap (with citation) or mark this axis PROCEED on that point.\n3. **Trust the diff over your intuition.** If the diff shows the\n   behavior IS covered and your instinct says it should have been\n   flagged, your instinct is wrong. Code review is reading the code\n   that\'s there, not inventing what you thought should be there.\n\n## JSON output contract\n\nReturn ONLY a JSON object. REVISE and BLOCK verdicts MUST include at\nleast one entry in `findings` with a concrete `file:line` citation\nfrom the diff. PROCEED may have an empty `findings` array.\n\n{\n  "score": <float 0-10>,\n  "result": "PROCEED" | "REVISE" | "BLOCK",\n  "ac_coverage": {"1": "ADDRESSED" | "UNADDRESSED" | "PARTIALLY", "2": "...", ...},\n  "comments": "one paragraph of concrete feedback about the DIFF",\n  "findings": [\n    {\n      "file":         "lib/path/to/file.ext",\n      "line":         <int or null: the post-change line number from\n                      the diff\'s `+` markers. Use null if the gap is\n                      about something NOT present in the diff>,\n      "severity":     "blocker" | "major" | "minor",\n      "description":  "one-sentence specific gap — what\'s wrong",\n      "evidence":     "1-3 quoted lines FROM the diff text OR\n                      \'absent: <what should be there>\' if the gap\n                      is about missing content"\n    }\n  ]\n}\n\n**Rules for findings:**\n- A REVISE verdict with an empty `findings` array is malformed. The\n  aggregator will coerce such a verdict to PROCEED because the\n  reviewer couldn\'t back up the claim with evidence.\n- `evidence` MUST quote actual diff lines or explicitly say\n  `absent: <description>`. If you can\'t quote a line AND can\'t\n  articulate what\'s absent, the concern probably doesn\'t exist —\n  drop it.\n- `line` refers to the post-change line number (the one after `+`).\n  If the gap is about missing content (nothing was added), use\n  `null` and put `absent: ...` in `evidence`.\n- `severity`: `blocker` for real bugs / security / correctness\n  violations; `major` for plan-fidelity / AC-gap / design issues;\n  `minor` for style / naming / nits.\n\nNo prose outside the JSON.\n]\n\nCard title: [BENCH t01 gemini-3-flash-preview-c] feat(tags): Add common_tags() helper for standardized resource tagging\nEvaluate the PR diff (shown in your system prompt) against the card + approved plan + any typed policy links.  Return the JSON verdict per the contract.', '-Q', '--yolo', '--max-turns', '5']' timed out after 300 seconds]

</details>

<details>
<summary>✅ <strong>code_quality</strong> — 7.90/10</summary>

The code itself is clean, proportionate, and well-placed for a small helper: the new module is a sensible home for the function, naming is clear, and the tests are focused and readable. My only blocker to merge is mechanical rather than craftsmanship-related: the diff shows the expected tests were added, but it does not provide evidence that the verification command was actually run and passed, so AC 2 is not fully closed from the review artifact provided.

</details>

<details>
<summary>✅ <strong>plan_fidelity</strong> — 9.70/10</summary>

The diff tracks the approved plan closely: it adds exactly the two planned files, implements `common_tags(env, team, project)` as a small standalone helper in `infiquetra_aws_infra/tagging.py`, strips all three inputs before validation, validates the normalized environment against the approved `dev/staging/prod` set, hard-codes `ManagedBy` to `CDK`, and adds the four planned pytest cases in `tests/test_tagging.py` for valid environments, invalid env rejection, whitespace normalization, and exact required keys. I see no mechanism drift, no silent scope expansion, and no extra abstractions or refactors beyond what the plan authorized.

</details>

<details>
<summary>✅ <strong>robustness</strong> — 8.60/10</summary>

From a robustness perspective this is merge-ready. The helper is a pure function with no I/O, shared state, or resource lifecycle risk, and its failure mode for unsupported environments is deterministic and explicit via ValueError after normalization. The whitespace normalization happens before validation, which avoids a common edge-case bug around padded env strings. I do note that non-str inputs would currently fail with AttributeError at .strip(), but given the typed str contract and the task scope, that is a minor hardening opportunity rather than a merge blocker.

</details>

<details>
<summary>✅ <strong>security_code</strong> — 8.60/10</summary>

This diff is a small pure-Python helper plus focused tests, and from a security-code perspective it does not introduce a meaningful attack surface. I verified there are no hardcoded secrets, no external calls, no dynamic execution, no shell/query/template construction, and no new unauthenticated or data-exposure path. The only input handling added is bounded normalization via `.strip()` and allowlist validation of `env` before returning a static dictionary, which is a safe pattern for the touched code path.

</details>

<details>
<summary>✅ <strong>python_types</strong> — 9.00/10</summary>

From a Python typing perspective, this diff is merge-ready. The new exported helper in infiquetra_aws_infra/tagging.py is fully annotated with concrete parameter and return types, the container return type is properly parameterized as dict[str, str], and the implementation matches that declared shape on every return path. The new pytest test functions are also annotated with -> None and use typed parametrization idiomatically. I do not see any introduced Any, untyped public API surface, or declared-type mismatch in the Python added by this PR.

</details>

<details>
<summary>✅ <strong>test_presence</strong> — 9.60/10</summary>

The diff materializes every test the card/plan called for in tests/test_tagging.py, and each test contains meaningful assertions tied to the observable ACs: valid dev/staging/prod behavior via parametrization, invalid env raising ValueError, whitespace normalization, and exact required-key coverage. The verification command listed in the card, pytest tests/test_tagging.py from the repo root, would exercise these new tests directly. No extra test surface or dependency changes appear in the diff, so from the test-presence lens this is merge-ready.

</details>

**Comment by namredips** [comment]
✅ Benchmark experiment reached merge-ready consensus. Closing without merging to keep `main` pristine for subsequent matrix runs. Scores live in the orchestrator SQLite `pr_reviews` table; artifacts are preserved in the branch and this PR history.
